# Data Scrubbing and Cleaning 

* Cast columns to appropriate data type
* Identify and deal with null values 
* Remove any unnecessary columns

Load Dataset and view data types and see if they correspond to the column descriptions.

In [1]:
import pandas as pd

data = pd.read_csv("data/data.csv")

In [2]:
for col in data:
    print(data[col].value_counts())

1699    1
1116    1
1120    1
1122    1
1124    1
       ..
573     1
575     1
577     1
579     1
2       1
Name: ID, Length: 1700, dtype: int64
63    90
65    81
62    79
64    68
70    66
      ..
27     2
92     2
36     2
30     1
26     1
Name: AGE, Length: 63, dtype: int64
1    1065
0     635
Name: SEX, dtype: int64
0    1060
1     410
2     147
3      79
?       4
Name: INF_ANAM, dtype: int64
0    661
6    332
1    146
2    137
5    125
3    117
?    106
4     76
Name: STENOK_AN, dtype: int64
2    854
0    661
?     73
3     54
1     47
4     11
Name: FK_STENOK, dtype: int64
2    683
1    548
0    418
?     51
Name: IBS_POST, dtype: int64
?    1628
0      45
1      27
Name: IBS_NASL, dtype: int64
2    880
0    605
3    195
1     11
?      9
Name: GB, dtype: int64
0    1635
1      57
?       8
Name: SIM_GIPERT, dtype: int64
0    551
7    432
?    248
6    165
1     93
5     73
3     58
2     58
4     22
Name: DLIT_AG, dtype: int64
0    1468
1     103
?      54
3      29
2      

# Cast Columns to Correct Data types

* SEX has no missing values and ranges from 0-1 
    * Nominal col, dtype=64 
    * Handling method is to remove rows with missing values
* INF_ANAM ranges from numbers 0-3 or Missing. Missing values are labeled '?' 4 missing values. 
    * Ordinal col, dtype=int64
    * Handling method is to remove rows with missing values
* STENOK_AN ranges from 0-6 it has 106 missing values labeled as '?'.   
    * Ordinal col, dtype=int64
    * Handling method is to remove rows with missing values
* FK_STENOK ranges from 0-4 but has 73 missing values labeled as '?'. 
    * Oridinal col, dtype=int64
    * Handling method is to remove rows with missing values
* IBS_POST ranges from 0-2 but has 51 missing values labeled as '?'. 
    * Ordinal col, dtype=int64
    * Handling method is to remove rows with missing values
* IBS_NASL ranges from 0-1, No, Yes, but has 9 missing values labeled as '?'. 
    * Nominal col, dtype=int64
    * Handling method is to remove rows with missing values
* GB ranges from 0-1, No, Yes, but has 1628 missing values labeled as '?'. 
    * Oridinal col, dtype=int64
    * Handling method is to remove rows with missing values
* SIM_GIPERT ranges from 0-1, No, Yes, but has 8 missing values labeled as '?'. 
    * Nominal col, dtype=int64 
    * Handling method is to remove rows with missing values
* DLIT_AG ranges from 0-7 but has 248 missing values labeled as '?'. 
    * Ordinal col, dtype=int64
    * Handling method is to remove rows with missing values 
* ZSN_A ranges from 0-4 but has 54 missing values labeled as '?'. 
    * Ordinal col, dtype=int64 
    * Handling method is to remove rows with missing values
* nr_11, , nr_01, nr_02, nr_03, nr_04, nr_07, nr_08 have ranges from 0-1, No, Yes, and all these columns have 21 missing values labeled as '?'. 
    * Nominal col
    * Handling method is to remove rows with missing values
* np_01, np_04, np_05, np_07, np_08, np_09, np_10 have ranges from 0-1, No, Yes, and all of these columns have 18 missig values labeled as '?'
    * Nominal col
    * Handling method is to remove rows with missing values
* endocr_01 has a range from 0-1, No, Yes, and has 11 missing values labeled as '?'
    * Nominal col, dtype=int64
    * Handling method is to remove rows with missing values
* endocr_02, endocr_03 have a range from 0-1, No, Yes, and have 10 missing values labeled as '?'.
    * Nominal col, dtype=int64
    * Handling method is to remove rows with missing values
* zab_leg_01, zab_leg_02, zab_leg_03, zab_leg_04, zab_leg_05, zab_leg_06, have a range from 0-1, 0=No, 1=Yes, and all have 7 missing values labeld as '?'. 
    * Nominal col, dtype=int64
    * Handling method is to remove rows with missing values
* S_AD_KBRIG has systolic blood pressures from the Emergency Cardiology Team  and has 1076 missing values labeled as '?' 
    * dtype=int64
    * Handling method is to remove rows with missing values
* D_AD_KBRIG diastolic blood pressure from the Emergency Cardiology Team  and has 267 missing values labeled as '?'. 
    * dtype=int64
    * Handling method is to remove rows with missing values
* S_AD_ORIT systolic blood pressure from the intensive care unit and has 267 missing values labeled as '?'.
    * dtype=int64
    * Handling method is to remove rows with missing values
* D_AD_ORIT diastolic blood pressure from the intensive care unit and has 267 missing values labeled as '?'.
    * dtype=int64
    * Handling method is to remove rows with missing values
* O_L_POST, Pulmonary edema at the time of admission to intensive care unit. Has 12 missing values labeled as '?'. 0 = No, 1 = Yes
    * Nominal col 
    * Handling method is to remove rows with missing 


There rest of the column are all approiate data types they all just have missing values and are either nominal or original. Since this is the case we can't simply impute a value since ranking can matter for some of these columns. 

According to the UCI repository there all multiple ways for multiclass predictions: 

1. the time of admission to hospital: all input columns (2-112) except 93, 94, 95, 100, 101, 102, 103, 104, 105 can be used for prediction;

2. the end of the first day (24 hours after admission to the hospital): all input columns (2-112) except 94, 95, 101, 102, 104, 105 can be used for prediction;

3.	the end of the second day (48 hours after admission to the hospital) all input columns (2-112) except 95, 102, 105 can be used for prediction;

4.	the end of the third day (72 hours after admission to the hospital) all input columns (2-112) can be used for prediction.

There is no need for casting columns to appropriate types since there all already in appropriate dtypes.


# Handle Missing Values and Neccesary Columns

We know that we have a lot of missing values in every column. We also know that we can not impute values in these columns since there categorical values that contain knowledge for that specific subject. We can't simply subsitute the missing values as that would impact our analysis and prediction models. The approach should be to drop the missing values row wise

Before we remove any columns we should consider the approach to the problem. The [UCI Machine Learning Repo](https://archive.ics.uci.edu/dataset/579/myocardial+infarction+complications) reccommends the following prediction problems to solve 

There are four possible time moments for complication prediction: on base of the information known at
1.	The time of admission to hospital: all input columns (2-112) except 93, 94, 95, 100, 101, 102, 103, 104, 105 can be used for prediction;
2.	The end of the first day (24 hours after admission to the hospital): all input columns (2-112) except 94, 95, 101, 102, 104, 105 can be used for prediction;
3.	The end of the second day (48 hours after admission to the hospital) all input columns (2-112) except 95, 102, 105 can be used for prediction;
4.	The end of the third day (72 hours after admission to the hospital) all input columns (2-112) can be used for prediction.

We can develop different datasets to approach each problem or we can attack a single problem. For sake of simplicity and time I will only be attempting one problem which is number 1. The main reason being I feel its a good area to start because generally people who have heart attacks have chest pain, shortness of and other symptoms a few days or weeks before the attack occurs. Most people would only be interested to enter a hospital when its very bad or too late and end up in the Emergency Room. This can be 1 or 2 days before an attack occurs and they just barely made it to the hospital to have personell ready for action.

[Source](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6307335/)

Handle the reccomended columns

In [3]:
subset = [
    "R_AB_1_n",
    "R_AB_2_n", 
    "R_AB_3_n", 
    "NA_R_1_n", 
    "NA_R_2_n", 
    "NA_R_3_n", 
    "NOT_NA_1_n", 
    "NOT_NA_2_n", 
    "NOT_NA_3_n"
]
data.drop(columns=subset, inplace=True)

## Handle null values

In [4]:
# Check null values
null = data.isnull().sum() > 0
null

ID           False
AGE          False
SEX          False
INF_ANAM     False
STENOK_AN    False
             ...  
DRESSLER     False
ZSN          False
REC_IM       False
P_IM_STEN    False
LET_IS       False
Length: 114, dtype: bool

There are no null values in the dataset. In reality the '?' values are our missing values. Lets see how many missing values we have in the dataset

In [5]:
# Replace '?' with NaN
data.replace('?', pd.NA, inplace=True)

a = data.isnull().sum()

# Calculate the percentage of missing values in each column
missing_percentage = (data.isnull().sum() / len(data)) * 100

# Select columns with missing values greater than 30%
columns_with_high_missing = missing_percentage[missing_percentage > 30]

# Print the columns with more than 30% missing values
print(columns_with_high_missing)


IBS_NASL      95.764706
S_AD_KBRIG    63.294118
D_AD_KBRIG    63.294118
KFK_BLOOD     99.764706
NA_KB         38.647059
NOT_NA_KB     40.352941
LID_KB        39.823529
dtype: float64


Its best to drop these columns entirely since they would mean removing about 500 rows or more from our dataset. Its worth droping these features if its means saving the rest of the columns:
* IBS_NASL
    * Identifies hereditary burden. 
* S_AD_KBRIG
    * Emergency room Systolic Pressure. It makes sense that most of these values are missing. It is a high pressure environment so logging every detail is probably not going to happen. 
* D_AD_KBRIG
    * Emergency room Diastolic Pressure. It makes sense that most of these values are missing. It is a high pressure environment so logging every detail is probably not going to happen. 
* KFK_BLOOD
    * Serum CPK contentm Creatine Phosphokinase. Group of enzymes found in the brain, heart, and skeletal muscle. Damage to these tissues can cause CPK to be released into the serum, which raises serum levels. 
    * This column is missing pretty much every subject.
* NA_KB 
    * Use of opioid drugs by the Emergency Cardiology Team.
    * Again its makes sense that this was failed to be logged in an emergency. Its a high pressure environment. 
* NOT_NA_KB
    * Use of NSAIDs by the Emergency Cardiology Team 
    * Again its makes sense that this was failed to be logged in an emergency. Its a high pressure environment.
* LID_KB
    * Use of lidocaine by the Emergency Cardiology Team. 
    * Again it makes sense that this was failed to be logged in an emergency. Its a high pressure environment.

In [6]:
# Drop rows containing NaN in the subset columns
subset = columns_with_high_missing
data.drop(columns=subset.index, inplace=True)

In [7]:
data.shape

(1700, 107)

Now lets take a look at the lower percentages in the columns

In [8]:
# Select columns with missing values between 10% and 30%
columns_with_medium_missing = missing_percentage[(missing_percentage > 10) & (missing_percentage <= 30)]

# Print the columns with missing values in the range of 10% to 30%
for col, percentage in columns_with_medium_missing.iteritems():
    print(f"{col}: {percentage:.2f}%")

DLIT_AG: 14.59%
S_AD_ORIT: 15.71%
D_AD_ORIT: 15.71%
GIPO_K: 21.71%
K_BLOOD: 21.82%
GIPER_NA: 22.06%
NA_BLOOD: 22.06%
ALT_BLOOD: 16.71%
AST_BLOOD: 16.76%
ROE: 11.94%


Highest Group:
* GIPO_K: 21.71%
* K_BLOOD: 21.82%
* GIPER_NA: 22.06%
* NA_BLOOD: 22.06%
    * The following columns are the highest in missing values. These values however are important. The estimation of sodium and potassium levels in acute Myocardial Infractions can help assess their prognosis according to a [journal article](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4290227/#:~:text=Hyponatremia%20and%20hypokalemia%20are%20indicators,was%20indicative%20of%20clinical%20improvement) published in PUBMED. 
    * That being said GIPO_K assesses Hypokalemia which is low potassium levels. GIPER_NA assesses high sodium levels in blood. We do not need the individual real time serum levels k_blood or na_bloos since we can account for them already.  
    * There is a huge ammount of data missing which can impact the regression. Normally I would not want to impute the data but because its about 20% of missing values thats about 350 rows. 
    * Since this data is nominal GIPO_K and GIPER_NA imputing the mode can still preserve the prevalance. 


Middle Group: 
* ALT_BLOOD: 16.71%
* AST_BLOOD: 16.76%
    * Alanine transaminase (ALT) and aspartate aminotransferase (AST) are referred to as liver transaminases. According to a [journal article](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4443194/) the results showed the serum levels of ALT and AST in Coronay Heart Disease (CHD) group were markedly higher than those in control group. Overall serum levels of ALT and AST could be used as independent predictors of CHD. There has been controversy of other studies where this was not the case and ALT and AST did not show as good predictors of CHD
    * Since this data is real time data I can not impute a value. It makes up about 15% of the data. It might be best to drop these columns since scientific research is still not showing confident results.

Lower Group: 
* DLIT_AG: 14.59%
    * Hypertension is still relevant for assesing coronary syndromes. This should stay. 
    * Impute the median so we still consider the ranking  
* S_AD_ORIT: 15.71%
    * Systolic blood pressure according to Emergency Cardiology Team
* D_AD_ORIT: 15.71%
    * Diastolic blood pressure according to Emergency Cardiology Team

    The issue with these values is that there related. DLIT_AG does account for BP since its an evaluation of the length of hypertension the person has had whether that be years or months. While S_AD and D_AD are real time values when they were assesed. Making use of S_AD and D_AD requires a professional evalution like a doctor saying its hypertension. Alone these columns aren't useful for a model unless they were interpreted which they were not. In that case we should keep DLIT_AG just so we account for Blood Pressure evaluation and drop the real time BP values (S_AD_ORIT, D_AD_ORIT).

* ROE: 11.94%
    * ESR (Erythrocyte sedimentation rate) 
    * According to a [journal article](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3333472/) Erythrocyte sedimentation rate may be a useful additional diagnostic criterion for coronary heart disease
    * Since this is a real time value I can not impute values. Keeping this column can mean sacrificing 200 rows of insight 

Impute the data

In [9]:
# Replace missing values in nominal columns with the mode
nominal_columns = ["GIPO_K", "GIPER_NA"]
data[nominal_columns] = data[nominal_columns].fillna(data.mode().iloc[0])

# Replace missing values in ordinal column with the median
ordinal_column = "DLIT_AG"
data[ordinal_column] = data[ordinal_column].fillna(data[ordinal_column].median())


Drop the rows with missing values for this low half of missing values

In [10]:
# Drop columns we decided to drop
subset = ["K_BLOOD", "NA_BLOOD", "S_AD_ORIT","D_AD_ORIT"]
data.drop(columns=subset, inplace=True)

In [11]:
data.shape

(1700, 103)

Lets take a look at the very low percentage values below 10 percent of missing values

In [215]:
# Select columns with missing values between 10% and 30%
columns_with_low_missing = missing_percentage[(missing_percentage < 10)]

# Sort the columns by missing value percentage in descending order
columns = columns_with_low_missing.sort_values(ascending=False)

# Print the columns with missing values in the range of 10% to 30%
for col, percentage in columns.iteritems():
    print(f"{col}: {percentage:.2f}%")

ritm_ecg_p_08: 8.94%
ritm_ecg_p_01: 8.94%
ritm_ecg_p_02: 8.94%
ritm_ecg_p_04: 8.94%
ritm_ecg_p_06: 8.94%
ritm_ecg_p_07: 8.94%
TIME_B_S: 7.41%
L_BLOOD: 7.35%
n_p_ecg_p_03: 6.76%
n_p_ecg_p_10: 6.76%
n_p_ecg_p_09: 6.76%
n_p_ecg_p_08: 6.76%
n_p_ecg_p_07: 6.76%
n_p_ecg_p_06: 6.76%
n_p_ecg_p_05: 6.76%
n_p_ecg_p_04: 6.76%
n_r_ecg_p_09: 6.76%
n_p_ecg_p_01: 6.76%
n_r_ecg_p_10: 6.76%
n_p_ecg_p_12: 6.76%
n_r_ecg_p_08: 6.76%
n_r_ecg_p_05: 6.76%
n_r_ecg_p_04: 6.76%
n_r_ecg_p_03: 6.76%
n_r_ecg_p_02: 6.76%
n_r_ecg_p_01: 6.76%
n_p_ecg_p_11: 6.76%
n_r_ecg_p_06: 6.76%
STENOK_AN: 6.24%
ant_im: 4.88%
inf_im: 4.71%
lat_im: 4.71%
FK_STENOK: 4.29%
post_im: 4.24%
ZSN_A: 3.18%
IBS_POST: 3.00%
nr_01: 1.24%
nr_03: 1.24%
nr_04: 1.24%
nr_07: 1.24%
nr_08: 1.24%
nr_11: 1.24%
nr_02: 1.24%
np_01: 1.06%
np_04: 1.06%
np_05: 1.06%
np_07: 1.06%
np_08: 1.06%
np_09: 1.06%
np_10: 1.06%
ASP_S_n: 1.00%
GEPAR_S_n: 1.00%
TRENT_S_n: 0.94%
TIKL_S_n: 0.94%
K_SH_POST: 0.88%
MP_TP_POST: 0.82%
ANT_CA_S_n: 0.76%
GT_POST: 0.71%
SVT_POST

The columns that are 'ritm' are specific to heart ECG diagnosis. However there seems to be a bias in these columns so imputing values can increase the bias but not by much since its aloready dominant and these columns are missing 150 rows which is a small fraction of the whole dataset and current size of the data we have which is about 950 rows with the current cleaning we have done. 
* Impute the mode for these nominal columns

Time elapsed from the beginning of the attack of CHD to the hospital (TIME_B_S): Ordinal column is the second largest column that is missing data from this subset. 
* We can impute the median 

White blood cell count (billions per liter) (L_BLOOD): Real time data. Since this is real time data we can't impute a value. This column however is important in detection of heart failure and possible outcomes of such an event. Increase occurs due to heart tissue damage. 
* Keep col and drop rows

All the columns that are 'n_p_ecg' are attributes of the ecg and are all nominal. * Impute the mode for these nominal columns

Exertional angina pectoris in the anamnesis (STENOK_AN): Ordinal col. 
Functional class (FC) of angina pectoris in the last year (FK_STENOK): Ordinal col. Classifies severity chest pain which can be heart disease. 
* Impute median

Presence of an anterior myocardial infarction (left ventricular) (ECG changes in leads V1: V4 ) (ant_im): Ordinal col. Presence of a lateral myocardial infarction (left ventricular) (ECG changes in leads V5: V6 , I, AVL) (lat_im): Ordinal col. Presence of an inferior myocardial infarction (left ventricular) (ECG changes in leads III, AVF, II). (inf_im): Ordinal. Presence of a posterior myocardial infarction (left ventricular) (ECG changes in V7: V9, reciprocity changes in leads V1 – V3) (post_im): Ordinal col. Again have to do with ecg interpretation and are important to predictions. 
* impute median 

Presence of chronic Heart failure (HF) in the anamnesis (ZSN_A): Partially ordered attribute: there are two lines of severities: 0<1<2<4, 0<1<3<4. Ordinal col. The conclusion of which side of ventricular valve did the heart fail. Can be useful for classifying the severity of heart failure from required predictors. 
* Impute median

In [250]:
data.dropna(axis=0, inplace=True)

In [251]:
data.shape

(930, 103)